interpreter/llvm-project/clang/lib/AST/ASTContext.cpp

```diff
@@ -4597,13 +4597,11 @@ QualType ASTContext::getAttributedType(attr::Kind attrKind,
 /// Retrieve a substitution-result type.
 QualType
 ASTContext::getSubstTemplateTypeParmType(const TemplateTypeParmType *Parm,
-                                         QualType Replacement,
-                                         bool AllowNonCanonical /* = false */
-                                        ) const {
-  if (!AllowNonCanonical) {
-     assert(Replacement.isCanonical()
-            && "replacement types must always be canonical");
-  }
+                                         QualType Replacement) const {
+#if 0
+  assert(Replacement.isCanonical()
+         && "replacement types must always be canonical");
+#endif
```

Comment:
As the hack was always enabled, this assertion was almost always disabled - there was one call site that passed `AllowNonCanonical = false` and a couple that used the default parameter, but the most important ones passed `HackForDefaultTemplateArg::AllowNonCanonicalSubst()`. When blindly reverting, I found that this assertion fails from the code path from `areEqualTypes` so I propose to unconditionally disable this.

Now the question is, where do I put this in the monorepo? https://github.com/root-project/llvm-project/commit/660281c8941e3169077d3b9cbde1aeeb1fc66fa5 is currently part of the patches for Cling - shall I add it there? Do we even need this patch for Cling or shall I move this to the patches for ROOT (`ROOT-llvm13`)? @vgvassilev

Reply:
We do not these patch in cling. We use it mostly to make ROOT I/O optimizations.

Reply:
Ok, I moved the commit in question to the `ROOT-llvm13` branch and will merge once the `llvm-diff` workflow is happy (no need to run the other CI tests again).